ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+
+```{python}
+#| code-fold: false
+sequence_length = 48  # Length of each input sequence
+X = []
+y = []
+
+for i in range(len(all_notes) - sequence_length):
+    X.append(all_notes[i:i+sequence_length])
+    y.append(all_notes[i+sequence_length])
+
+X = np.array(X)
+y = np.array(y)
+print("Shape of X", X.shape)
+print("Shape of y", y.shape)
+```
+
+## LSTM Architecture
+
+We now create the LSTM architecture.
+
+
+```{python}
+#| code-fold: false
+import torch
+import torch.nn as nn
+
+class MusicLSTM(nn.Module):
+    def __init__(self, input_dim, hidden_dim, num_layers, output_dim):
+        super(MusicLSTM, self).__init__()
+        self.hidden_dim = hidden_dim
+        self.num_layers = num_layers
+        
+        self.lstm = nn.LSTM(input_dim, hidden_dim, num_layers, batch_first=True)
+        self.fc = nn.Linear(hidden_dim, output_dim)
+    
+    def forward(self, x):
+        h0 = torch.zeros(self.num_layers, x.size(0), self.hidden_dim).to(x.device)
+        c0 = torch.zeros(self.num_layers, x.size(0), self.hidden_dim).to(x.device)
+        
+        out, _ = self.lstm(x, (h0, c0))
+        out = self.fc(out[:, -1, :])
+        return out
+
+# Example usage
+input_dim = X.shape[2]  # Number of features (e.g., pitch, velocity)
+hidden_dim = 20
+num_layers = 2
+output_dim = X.shape[2]
+model = MusicLSTM(input_dim, hidden_dim, num_layers, output_dim)
+```
+
+
+## Training
+
+We will train our model for 10 epochs.
+
+```{python}
+criterion = nn.MSELoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+num_epochs = 100
+batch_size = 8
+for epoch in range(num_epochs):
+    for i in range(0, len(X), batch_size):
+        inputs = torch.tensor(X[i:i+batch_size], dtype=torch.float32)
+        targets = torch.tensor(y[i:i+batch_size], dtype=torch.float32)
+        
+        outputs = model(inputs)
+        loss = criterion(outputs, targets)
+        
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+```
+
+## Generate
+
+We next write a function that will help us generate a new piece of music.
+
+```{python}
+#| code-fold: false
+def generate_music(model, start_sequence, length):
+    model.eval()
+    generated = start_sequence
+    for _ in range(length):
+        input_seq = torch.tensor(generated[-sequence_length:], dtype=torch.float32).unsqueeze(0)
+        next_note = model(input_seq).detach().numpy()
+        generated = np.vstack((generated, next_note))
+    return generated
+
+# Example usage
+start_sequence = X[10]  # Starting sequence for generation
+generated_music = generate_music(model, start_sequence, 100)  # Generate 100 notes
+print(generated_music.shape)
+```
+
+## Download the Song
+
+We can generate an output midi file to play. The song is somewhat similar to the original piece. We could probably create a better song through more training and a larger training set.
```

Comment:
Again, could we add an html control to play the output? I can help you with that if you want.

Reply:
added in my PR to this branch, but the result is not good.

Reply:
Oh, actually, if the model is generating music samples, it would be better to just convert the generated music sample to a wav file and skip MIDI, no?

Reply:
Yes, I would need help with this.

Reply:
Ok. can you explain what you are generating? Is midi_to_notes converting to simply time sampled audio data? do you know what sampling frequency?

Reply:
The model is predicting note start, note end, note pitch and note velocity. This function should be taking this generated data and converting it to a midi file. I'm not sure what the sampling frequency is. I used copilot to help write this function and I'm not entirely sure what the prettymidi library is doing under the hood, aside from generating these .midi files. I would need to look this up in more detail.

Reply:
Oh. i see from prettyMIDI that it creates numpy arrays of start time, end time, pitch and velocity sequences... Is that what you intended to give as input to the RNN? I had just assumed you were converting regularly sampled audio data.

Reply:
Yes, I'm training the model on that data. I am adding more to the description in my slides.

Reply:
Got it. And I see this Transformer model is using a similar structure of the data too: https://magenta.tensorflow.org/maestro-wave2midi2wave